File: app/src/main/java/com/example/indroydlab/cart/CartManager.java
package com.example.indroydlab.cart;

import com.example.indroydlab.model.Cart_item;
import com.example.indroydlab.model.Product;
import java.util.ArrayList;
import java.util.List;

public class CartManager {
    private final List<Cart_item> cartItems = new ArrayList<>();

    public List<Cart_item> getCartItems() {
        return new ArrayList<>(cartItems);
    }
    public void addToCart(Product product) {
        for (Cart_item item : cartItems) {
            if (item.getProduct().getId() == product.getId()) {
                item.incrementQuantity();
                return;
            }
        }
        cartItems.add(new Cart_item(product));
    }

    public double getSubTotal() {
        double total = 0;
        for (Cart_item item : cartItems) {
            total += item.getItemTotal();
        }
        return total;
    }

    public double getTaxTotal() {
        double tax = 0;
        for (Cart_item item : cartItems) {
            tax += item.getItemTotal() * item.getProduct().getTaxPercent() / 100;
        }
        return tax;
    }


    public double getFinalAmount() {
        double total = getSubTotal() + getTaxTotal() - getCouponDiscount();
        return total;
    }


    public double getCouponDiscount() {
        double eligibleAmount = 0;

        for (Cart_item item : cartItems) {
            if (!item.getProduct().isDiscounted()) {
                eligibleAmount += item.getItemTotal();
            }
        }

        if (eligibleAmount < 1000) {
            return 0;
        }

        double discount = eligibleAmount * 0.20;

        return Math.min(discount, 300);
    }

    public void increaseQuantity(Product product) {
        for (Cart_item item : cartItems) {
            if (item.getProduct().getId() == product.getId()) {
                item.setQuantity(item.getQuantity() + 1);
                System.out.println("Qty = " + item.getQuantity());
                return;
            }
        }
    }

    public void decreaseQuantity(Product product) {
        for (int i = 0; i < cartItems.size(); i++) {
            Cart_item item = cartItems.get(i);
            if (item.getProduct().getId() == product.getId()) {
                if (item.getQuantity() > 1) {
                    item.setQuantity(item.getQuantity() - 1);
                } else {
                    cartItems.remove(i);
                }
                return;
            }
        }
    }

    public void clearCart() {
        cartItems.clear();

    }
    public void removeFromCart(Product product) {
        for (int i = 0; i < cartItems.size(); i++) {
            if (cartItems.get(i).getProduct().getId() == product.getId()) {
                cartItems.remove(i);
                return;
            }
        }
    }



}

File: app/src/main/java/com/example/indroydlab/model/CartManager.kt
package com.example.indroydlab.model

import androidx.compose.runtime.mutableStateListOf

object CartManager {
    val cartItems = mutableStateListOf<CartItem>()

    fun addToCart( product: ProductModel ) {
        val existingItem = cartItems.find { it.product.id == product.id }
        if (existingItem != null) {
            val index = cartItems.indexOf(existingItem)
            cartItems[index] = existingItem.copy(quantity = existingItem.quantity + 1)

        } else { cartItems.add(CartItem(product)) }
    }
    fun getSubTotal(): Double = cartItems.sumOf { it.getItemPrice() }
    fun getTaxTotal(): Double = cartItems.sumOf { it.getTaxAmount() }
    fun getCouponDiscount(): Double {
        val eligibleAmount = cartItems
            .filter { it.product.discountedPrice == null }
            .sumOf { it.getItemPrice() }
        if (eligibleAmount < 1000) return 0.0
        return minOf(eligibleAmount * 0.20,300.0)
    }
    fun getFinalAmount(): Double = getSubTotal() + getTaxTotal() - getCouponDiscount()
    fun increaseQuantity(product: ProductModel) {
        val index = cartItems.indexOfFirst { it.product.id == product.id }
        if (index != -1) {
            val item = cartItems[index]
            cartItems[index] = item.copy(quantity = item.quantity + 1)
        }
    }
    fun decreaseQuantity(product: ProductModel) {
        val item = cartItems.find { it.product.id == product.id }
        item?.let {
            if (it.quantity > 1){
                it.quantity--
            }else{
                cartItems.remove(it)
            }
        }
    }
    fun removeFromCart(product: ProductModel) =
        cartItems.removeAll { it.product.id == product.id }
    fun clearCart() = cartItems.clear()
}